src/modm/platform/dac/stm32/dac.hpp.in

```diff
@@ -0,0 +1,224 @@
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+// ----------------------------------------------------------------------------
+
+#ifndef MODM_STM32_DAC{{ id }}_HPP
+#define MODM_STM32_DAC{{ id }}_HPP
+
+#include <cstdint>
+#include "../device.hpp"
+#include <modm/platform/gpio/connector.hpp>
+#include <modm/platform/clock/rcc.hpp>
+#include <modm/src/modm/math/units.hpp>
+
+namespace modm::platform
+{
+
+/**
+ * Digital-to-Analog Converter (DAC{{ id }})
+ *
+ * Supports simple synchronous access to the 2-channel DAC found on most STM32
+ * microcontrollers.
+ *
+ * @author Jeff McBride
+ * @ingroup modm_platform_dac{% if id %} modm_platform_dac_{{id}}{% endif %}
+ */
+class Dac{{ id }} {
+public:
+	/**
+%% if dual_channel
+	 * The DAC has two output channels, Channel1 and Channel2
+%% else
+	 * The DAC has one output channel, Channel1
+%% endif
+	 */
+	enum class Channel : uint8_t
+	{
+		Channel1 = 0,
+%% if dual_channel
+		Channel2 = 1
+%% endif
+	};
+
+%% if has_mcr
+	enum class Mode
+	{
+		ExternalWithBuffer = 0,
+		ExternalInternalWithBuffer = DAC_MCR_MODE1_0,
+		ExternalWithoutBuffer = DAC_MCR_MODE1_1,
+		Internal = DAC_MCR_MODE1_0 | DAC_MCR_MODE1_1
+	};
+%% endif
+
+	template< template<Peripheral _> class... Signals >
+	static void
+	connect()
+	{
+		using Connector = GpioConnector<Peripheral::Dac{{ id }}, Signals...>;
+		Connector::connect();
+	}
+
+%% if target.family == "g4"
+	/**
+	 * Initialize the D/A converter.
+	 *
+	 * Enables the RCC clock output for the DAC
+	 *
+	 * @tparam SystemClock System clock struct
+	 */
+	template<typename SystemClock>
+	static inline void initialize();
+
+%% else
+	/**
+	 * Initialize the D/A converter.
+	 *
+	 * Enables the RCC clock output for the DAC
+	 */
+	static void initialize();
+
+%% endif
+
+	/** Enable DAC output for the given channel
+	 *
+	 * @param chan    The channel to be enabled
+	 *
+	 * @pre The DAC clock must be enabled with initialize()
+	 */
+	static void enableChannel(Channel chan);
+
+	/**
+	 * Disable DAC output for the given channel
+	 *
+	 * @param chan    The channel to be disabled
+	 *
+	 * @pre The DAC clock must be enabled with initialize()
+	 */
+	static void disableChannel(Channel chan);
+
+%% if has_mcr
+	/**
+	 * Control the output mode
+	 *
+	 * @param chan  The channel to configure the mode for
+	 * @param mode  The channel output mode
+	 *
+	 * @warning The DAC channel will be disabled when setting the mode
+	 */
+	static void setMode(Channel chan, Mode mode);
+
+%% else
+	/**
+	 * Control the output buffer setting
+	 *
+	 * @param chan    The channel to enable/disable the output buffer on
+	 * @param enable  true to turn on the output buffer, false to turn it off
+	 *
+	 * @pre The DAC clock must be enabled with initialize()
+	 */
+	static void enableOutputBuffer(Channel chan, bool enable);
+
+%% endif
```

Comment:
How about using the function signature `setMode(...)` instead of the `enableOutputBuffer(...)` function also for devices without MCR with `enum class Mode { ExternalWithBuffer, ExternalWithoutBuffer};`?
This would probably make code more portable in many places...

Reply:
I'd rather prefer to not break existing user code. What do you think about adding `setMode()` for all STM32 variants and only generating but deprecating `enableOutputBuffer()` for F4? For the DMA version we can do what we want since it's introduced here.

Reply:
Valid point! While vieving the diff I overlooked the old code (because the files were renamed).

We maybe could keep the old API with a deprecation warning and implement the new more flexibel API for all devices, but that is not important (to me).